Лабораторная работа 1 готова к проверке.

inittose opened the pull request:
## Изменения в репозитории:
- Переименовал ветку lab1 -> feature/lab1
- создал develop, куда буду заливать коммиты с feature/..

inittose committed:
Реализовал класс Item

inittose committed:
Создал класс Customer и класс ValueValidator

inittose committed:
Сверстал пользовательский интерфейс управления ItemsTab

inittose committed:
Возобновил настройки решения

inittose committed:
Начало реализации логики выбора/редактирования товаров

inittose committed:
Рабочая логика, но без комментариев и возможен рефакторинг

gag410 committed:
Добавил события изменения полей TextBox в ItemsTab

gag410 committed:
Добавил иконку и переименовал MainForm

inittose committed:
Рефакторинг логики добавления/редактирования товаров в ItemsTab

inittose committed:
Сверстал CustomersTab

inittose committed:
Изменение в верстках CustomersTab и ItemsTab

inittose committed:
Реализовал логику для CustomersTab

inittose committed:
Создал UML-диаграмму, добавил в readme

inittose committed:
Поменял иконку + удалил лишние файлы в решении проекта

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Customer.cs:
```diff
@@ -0,0 +1,75 @@
+        /// </summary>
+        public int Id
+        {
+            get
```
Для аксессоров (геттеров, сеттеров, тд), записанных в одну строку, используй запись в виде выражения get => _value;
Исправь во всём решении

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Item.cs:
```diff
@@ -0,0 +1,132 @@
+            }
+            set
+            {
+                if (value < 0 || value > 100000f)
```
1. Перенести эту проверку в класс валидатора. У тебя уже есть класс валидатора, который выполняет валидацию значений (например, по макс. длине), это значит, что подобные проверки следует делегировать ему. Потому что этот код потом можно переиспользовать, а значит уменьшить его дублирование и сократить время на написание новой функциональности
2. Любые магические числа следует выносить в отдельные переменные, это облегчает чтение кода как тебе, так и другим

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Customer.cs:
```diff
@@ -0,0 +1,75 @@
+    /// <summary>
+    /// Хранит данные о покупателе.
+    /// </summary>
+    internal class Customer
```
Модификатор internal позволяет использовать этот класс только внутри данной сборки (проекта). В контексте лабораторных по этому предмету это не имеет значения, однако ты должен понимать, в чём значение этого модификатора, если его используешь. Если этот класс нужно будет покрыть юнит-тестами (которые описываются в другом проекте), то этого сделать не получится без изменения модификатора доступа на public. Подробнее о модификаторах доступа читать вот тут: https://metanit.com/sharp/tutorial/3.2.php

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Item.cs:
```diff
@@ -0,0 +1,132 @@
+            {
+                if (value < 0 || value > 100000f)
+                {
+                    throw new ArgumentException("Cost is not included in the range from 0 to 100 000.");
```
Тут сообщение на английском, а в валидаторе - на русском. Сделать единообразно

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/IdGenerator.cs:
```diff
@@ -0,0 +1,28 @@
+        /// Возвращает уникальный идентификатор и вычисляет новый.
+        /// </summary>
+        /// <returns>Текущий уникальный идентификатор.</returns>
+        static public int GetNextId()
```
Сначала указывается модификатор доступа, потом модификатор static (см. стандарт оформления кода RSDN)

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+        /// <param name="selectedIndex">Индекс товара в списке.</param>
+        private void SetTextBoxes(int selectedIndex)
+        {
+            bool status = selectedIndex >= 0;
```
По названию переменной status не понятно, за что она отвечает. Для флагов лучше использовать стиль именования isSomething, например, isSelectedIndexCorrect. Переименовать

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+        private void RemoveButton_Click(object sender, EventArgs e)
+        {
+            int removeIndex = CustomersListBox.SelectedIndex;
+            if (removeIndex != -1)
```
Обращай внимание на вложенность метода. Чем больше вложенность, тем сложнее воспринимать код разработчику. В данном случае можно убрать лишний уровень вложенности инвертировав это условие:
if (removeIndex == -1)
{
return;
}

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+            {
+                CustomersListBox.Items.RemoveAt(removeIndex);
+                _customers.RemoveAt(removeIndex);
+                if (CustomersListBox.Items.Count > 0)
```
Инвертировать условие

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/ItemsTab.cs:
```diff
@@ -0,0 +1,217 @@
+        {
+            Item newItem = new Item();
+            newItem.Name = $"Item{newItem.Id}";
+            _items.Add(newItem);
```
Запрещено обращаться к приватному полю в обход аксессора. Обернуть _items в свойство или заменить его на авто-свойство.

inittose replied:
Правильно ли я понимаю, что свойство может быть как `private`, так и `public`? При этом стандарт написания имени свойства также будет использовать стиль `Паскаль` для регистра букв?

> Пример, что я имею ввиду: то есть к закрытому полю `_items` можно создать свойство как `public List<Item> Items`, так и `private List<Item> Items`

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/ItemsTab.cs:
```diff
@@ -0,0 +1,217 @@
+        private void RemoveButton_Click(object sender, EventArgs e)
+        {
+            int removeIndex = ItemsListBox.SelectedIndex;
+            if (removeIndex != -1)
```
Инвертировать

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/ItemsTab.cs:
```diff
@@ -0,0 +1,217 @@
+            {
+                ItemsListBox.Items.RemoveAt(removeIndex);
+                _items.RemoveAt(removeIndex);
+                if (ItemsListBox.Items.Count > 0)
```
Инвертировать

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/ItemsTab.cs:
```diff
@@ -0,0 +1,217 @@
+            {
+                WrongCostLabel.Text = "Cost must be greater than 0.";
+            }
+            else if (getParse > 100000)
```
Вынести "магические числа в переменную или константу

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/ItemsTab.cs:
```diff
@@ -0,0 +1,217 @@
+            {
+                WrongNameLabel.Text = "Name must consist of characters.";
+            }
+            else if (NameTextBox.Text.Length > 200)
```
"Магическое число"

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/ItemsTab.cs:
```diff
@@ -0,0 +1,217 @@
+                return;
+            }
+            Color currentColor = wrongInputColor;
+            if (DescriptionTextBox.Text.Length > 1000)
```
"Магическое число"

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/IdGenerator.cs:
```diff
@@ -0,0 +1,28 @@
+    /// <summary>
+    /// Хранит и возвращает уникальный идентификатор.
+    /// </summary>
+    static internal class IdGenerator
```
Сначала идёт модификатор доступа internal: internal static class
См. стандарт оформления кода RSDN: https://www.rsdn.org/article/mag/200401/codestyle.XML

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+
+namespace ObjectOrientedPractics.View.Tabs
+{
+    public partial class CustomersTab : UserControl
```
XML

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+        private Color rightInputColor = Color.White;
+        private Color wrongInputColor = Color.Red;
+
+        public CustomersTab()
```
XML. Исправить во всём решении

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+        /// </summary>
+        private List<Customer> _customers = new List<Customer>();
+
+        private Color rightInputColor = Color.White;
```
XML

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+        private List<Customer> _customers = new List<Customer>();
+
+        private Color rightInputColor = Color.White;
+        private Color wrongInputColor = Color.Red;
```
XML

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+        {
+            Customer newCustomer = new Customer();
+            newCustomer.FullName = $"Customer{newCustomer.Id}";
+            _customers.Add(newCustomer);
```
Обернуть в свойство

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+            {
+                WrongFullNameLabel.Text = "Full name must consist of characters.";
+            }
+            else if (FullNameTextBox.Text.Length > 200)
```
"Магические числа"
Исправить во всём решении

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,170 @@
+
+        private void RemoveButton_Click(object sender, EventArgs e)
+        {
+            int removeIndex = CustomersListBox.SelectedIndex;
```
Цитирую моего ревьювера "Старайся использовать var везде, где не нужно явно указывать тип при создании переменной, исключением является цикл for. На работу программы не влияет, есть аргументы за и против var, но такая политика компании."

Nikita-Isaichenko replied:
Цитирую моего ревьювера "Старайся использовать var везде, где не нужно явно указывать тип при создании переменной, исключением является цикл for. На работу программы не влияет, есть аргументы за и против var, но такая политика компании."

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/ItemsTab.cs:
```diff
@@ -0,0 +1,217 @@
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows.Forms;
```
Удалить не использующиеся using, также и в других файлах

Nikita-Isaichenko replied:
Удалить не использующиеся using, также и в других файлах

inittose committed:
Реализовал рандомную генерацию CustomerFactory и ItemFactory

inittose committed:
feature (lab1): Реализована сериализация проекта

- Создан статический класс Serializer
- Убрана возможность fullscreen в MainForm

ВАЖНО: Подключил к проекту Newtonsoft JSON .NET

inittose committed:
style(lab1): Внес правки от ревьюеров

- Сверился со стандартом оформления кода RSDN
- Инвертировал некоторые логические конструкции
- Заменил явное типизирование на var в методах
- Убрал магические числа
- Добавил комментарии, там где их не хватало
- Удалил неиспользующиеся using'и
- Дополнил и поправил UML-диаграмму

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Customer.cs:
```diff
@@ -0,0 +1,114 @@
+        /// <summary>
+        /// Максимальное число символов имени покупателя.
+        /// </summary>
+        private static readonly int _fullnameLengthLimit = 200;
```
Зачем static readonly, не проще ли использовать const?

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/IdGenerator.cs:
```diff
@@ -0,0 +1,31 @@
+        /// <summary>
+        /// Возвращает и задает счетчик уникальных идентификаторов.
+        /// </summary>
+        private static int Counter
```
Сделай автоматическим свойством

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/Serializer.cs:
```diff
@@ -0,0 +1,138 @@
+        /// <summary>
+        /// Возвращает путь до файла сериализации.
+        /// </summary>
+        public static String FilePath
```
Если в сеттере или геттере не производишь никаких действий с переменной, то лучше использовать авто свойства, это позволит уменьшить код

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,227 @@
+        /// <summary>
+        /// Список покупателей класса <see cref="Customer"/>
+        /// </summary>
+        private List<Customer> _customers = Serializer.GetCustomers();
```
Можно объявить как авто свойство и сразу в него присвоить. Посмотреть во всем коде

Nikita-Isaichenko commented on doc/Practics1.png:
Стрелки от CustomerFactory, ItemFactory, Serializer

inittose committed:
refactoring(lab1): Убрал лишние свойства

- Заменил некоторые поля на const
- Свойства без валидации заменил на авто-свойства

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/ItemFactory.cs:
```diff
@@ -12,38 +12,22 @@ public static class ItemFactory
         /// Путь до базы данных товаров.
         /// </summary>
-        private static readonly String _fileName =
+        private static String FileName { get; } =
```
Это не замечание, просто предложение добавить set, чтобы можно было например указывать другие пути к базам данным.

inittose replied:
Конечно можно, но в данном классе такой функционал пока не реализован.

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/ItemsTab.cs:
```diff
@@ -15,43 +15,27 @@ public partial class ItemsTab : UserControl
         /// Цвет <see cref="TextBox"/>, успешно прошедшего валидацию. 
         /// </summary>
-        private static readonly Color _rightInputColor = Color.White;
+        private static Color RightInputColor { get; } = Color.White;
```
А можешь объяснить зачем ты используешь static?

inittose replied:
Для данного класса бессмысленно, как я только что понял. У нас же только один объект данного класса будет использоваться, поэтому нет смысла для static.

Nikita-Isaichenko commented on doc/Practics1.png:
Ты указал стрелки для реализации, реализация применяется для обозначения классов, которые реализуют интерфейсы, здесь уже у тебя обычное использование.

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Item.cs:
```diff
@@ -0,0 +1,146 @@
+        /// </summary>
+        public Item()
+        {
+            Name = String.Empty;
```
String - класс из пространства имён System, а string - alias (псевдоним), за которым скрывается System.String. Поведение программы не меняется (только если его не переопределять для string), но используй string (первая - строчная)

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,203 @@
+            }
+        }
+
+        private void CustomersListBox_SelectedIndexChanged(object sender, EventArgs e)
```
XML. Исправить во всём решении

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CustomersTab.cs:
```diff
@@ -0,0 +1,203 @@
+
+        private void AddButton_Click(object sender, EventArgs e)
+        {
+            Customer newCustomer = new Customer();
```
Используй var вместо типов

inittose committed:
style (lab1): Добавил комментарии к событиям

- Также изменил стрелки в UML-диаграмме
- Заменил явную типизацию, на неявную (var)
- String -> string
